File: kotlin/0021-merge-two-sorted-lists.kt
class Solution {
    fun mergeTwoLists(l1: ListNode?, l2: ListNode?): ListNode? {
        var r1 = l1
        var r2 = l2
        val ll = ListNode(0)
        var rr = ll
        while(true){
            if(r1==null && r2==null){
                break
            }
            if(r1==null){
                rr.next = r2
                break
            }
            if(r2==null){
                rr.next = r1
                break
            }
            if(r1.`val`<=r2.`val`){
                rr.next = r1
                r1 = r1.next
            }
            else{
                rr.next = r2
                r2 = r2.next
            }
            rr = rr.next
        }
        return ll.next
    }
}

File: ruby/0021-merge-two-sorted-lists.rb
# Iterative solution
def merge_two_lists(list1, list2)
    dummy = ListNode.new
    node = dummy
  
    while list1 && list2
      if list1.val < list2.val
        node.next = list1
        list1 = list1.next
      else
        node.next = list2
        list2 = list2.next
      end
      node = node.next
    end
  
    node.next = list1 || list2
  
    dummy.next
end

# Recursive solution
def merge_two_lists(list1, list2)
    return list2 if list1.nil?
    return list1 if list2.nil?

    if list1.val <= list2.val
        list1.next = merge_two_lists(list1.next, list2)
        return list1
    else
        list2.next = merge_two_lists(list1, list2.next)
        return list2
    end
end
